File: testsuites/HLT/Benchmark/concurrency-benchmark/WaitNotifyExtended/loom/WaitNotifyBase.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

import java.util.concurrent.CountDownLatch;


public class WaitNotifyBase extends WaitNotifyBench {

  public WaitNotifyBase(int consumers, int producers, int valuePerProducer) {
    super(consumers, producers, valuePerProducer);
  }

  public String name() { return "WaitNotifyBase      "; }

  static Object monitorForConsumer = new Object();
  static Object monitorForProducer = new Object();

  static class Consumer extends WaitNotifyBench.Consumer {
    public Consumer(int id, WorkList workList) { super(id, workList); }

    protected void consumerLoop() throws InterruptedException {
      synchronized (monitorForConsumer) {
        while (workList.isEmpty()) {
          monitorForConsumer.wait();
        }

        Object work = workList.pop();
        consumed += 1;

        if (workList.isEmpty()) {
          synchronized (monitorForProducer) { monitorForProducer.notify(); }
        }
      }
    }
  }

  static class Producer extends WaitNotifyBench.Producer {
    public Producer(int id, WorkList workList, int valPerProducer,
                    CountDownLatch startSignal) {
      super(id, workList, valPerProducer, startSignal);
    }

    protected void producerLoop() throws InterruptedException {
      synchronized (monitorForConsumer) {
        for (int i = 0; i < VALUE_PER_PRODUCER; i++)
          workList.push(CONSUMING_UNIT);
        monitorForConsumer.notifyAll();
      }

      synchronized (monitorForProducer) {
        while (!workList.isEmpty()) {
          monitorForProducer.wait();
        }
      }
    }
  }

  public WaitNotifyBench.Consumer createConsumer(int id) {
    return new Consumer(id, workList);
  }

  public WaitNotifyBench.Producer createProducer(int id,
                                                 CountDownLatch startSignal) {
    return new Producer(id, workList, VALUE_PER_PRODUCER, startSignal);
  }
}

File: testsuites/HLT/Benchmark/concurrency-benchmark/WaitNotifyExtended/go/WaitNotifyBase.go
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

package main

import (
  "sync"
  "sync/atomic"
)

type base struct {
  muForConsumer *sync.Mutex
  coForConsumer *sync.Cond
  muForProducer *sync.Mutex
  coForProducer *sync.Cond
}

func NewBase() *base {
  var mc = sync.Mutex{}
  var mp = sync.Mutex{}

  return &base{muForConsumer: &mc,
    coForConsumer: sync.NewCond(&mc),
    muForProducer: &mp,
    coForProducer: sync.NewCond(&mp)}
}

func (b *base) name() string {
  return "base"
}

func (b *base) consumerLoop(c *Consumer) {
  defer b.coForConsumer.L.Unlock()
  b.coForConsumer.L.Lock()
  for c.workList.isEmpty() {
    if atomic.LoadInt64(&c.shouldStop) != 0 {
      return
    }
    b.coForConsumer.Wait()
  }

  c.workList.pop()
  c.consumed++

  if c.workList.isEmpty() {
    b.coForProducer.L.Lock()
    b.coForProducer.Signal()
    b.coForProducer.L.Unlock()
  }
}

func (b *base) producerLoop(p *Producer) {
  b.coForConsumer.L.Lock()
  for i := 0; i < p.VALUE_PER_PRODUCER; i++ {
    p.workList.push(CONSUMING_UNIT)
    b.coForConsumer.Broadcast()
  }
  b.coForConsumer.L.Unlock()

  b.coForProducer.L.Lock()
  for !p.workList.isEmpty() {
    if atomic.LoadInt64(&p.shouldStop) != 0 {
      b.coForProducer.L.Unlock()
      return
    }
    b.coForProducer.Wait()
  }
  b.coForProducer.L.Unlock()
}

func (b *base) start() {}

func (b *base) finish() {
  b.coForConsumer.L.Lock()
  b.coForConsumer.Broadcast()
  b.coForConsumer.L.Unlock()
  b.coForProducer.L.Lock()
  b.coForProducer.Broadcast()
  b.coForProducer.L.Unlock()
}

func (b *base) createConsumer(id int, endSignal *sync.WaitGroup) *Consumer {
  return &Consumer{id, &WORKLIST, 0, b.consumerLoop, endSignal, 0}
}

func (b *base) createProducer(id int, startSignal, endSignal *sync.WaitGroup) *Producer {
  return &Producer{id, &WORKLIST, VALUE_PER_PRODUCER, startSignal, b.producerLoop, endSignal, 0}
}
